src/music/MusicManager.ts

```diff
@@ -0,0 +1,213 @@
+
+    function scheduleNote(note: Note, inst: Instrument): void {
+        const instOut: AudioNode = inst.scheduleNote(context, note);
+        instOut.connect(masterGain);
+    }
+
+
+    function octave(root: number, octave: number): number {
+        return mod(root, 12) + octave * 12;
+    }
+
+    function regenerateState(): void {
+        // Allowed time signatures, as n/8
+        const timeSignatures: NonEmptyArray<number> = [5, 6, 7, 8, 9, 11, 13];
+        state.rhythm.generateNewSubdivision(Random.fromArray(timeSignatures));
+        // Query all greek modes (only one imperfection) with a perfect fifth
+        // above the tonic
+        const query: ScaleQuery = {
+            chord: 1 << 7,
+            imperfections: [1, 1]
+        };
+        const matchingScales = Scales.getAllScalesMatchingQuery(query);
+        if (Arrays.isNonEmpty(matchingScales)) {
+            state.scale = Random.fromArray(matchingScales);
+        }
+        state.drumLoop = generateDrumLoop();
+        state.progression = generateChordProgression(state.scale, state.root);
+        state.queue = [
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.CHORDS, MeasureContents.DRUMS],
+            [MeasureContents.DRUMS]
+        ];
+    }
+
+    function queueChord(startingTime: number): void {
+        const curChord = state.progression[mod(state.chordIndex, state.progression.length)];
+        const beatLength: number = 60 / state.beatsPerMinute;
+        for (const val of curChord) {
+            scheduleNote({
+                midiNumber: octave(val, 5),
+                start: startingTime,
+                duration: beatLength * state.rhythm.beats
+            }, instruments.pad);
+        }
+        state.chordIndex++;
+    }
+
+    function queueDrumLoop(startingTime: number): void {
+        const beatLength: number = 60 / state.beatsPerMinute;
+        let offsetTime = 0;
+        for (const drum of state.drumLoop) {
+            scheduleDrum(startingTime + offsetTime, drum);
+            offsetTime += beatLength;
+        }
+    }
+
+    function queueNextMeasures(startingTime: number): void {
+        const beatLength: number = 60 / state.beatsPerMinute;
+        if (state.queue.length === 0) {
+            regenerateState();
+        }
+        const measureLength: number = beatLength * state.rhythm.beats;
+        if (state.queue.length > 0) {
+            // this cast is safe since it's only performed if there is anything in the queue
+            const currentContents: MeasureContents[] = state.queue.shift() as MeasureContents[];
```

Comment:
You can just use `!` instead of casting here

Reply:
You can also write a utility function `safeShift<T>(arr: NonEmptyArray<T>): T` for more safety and then use `Arrays.isNonEmpty` instead, but then the argument would still be inferred to be nonempty after you shift, which would kind of defeat the safety.

Reply:
Actually, the best way would just be to `const currentContents = state.queue.shift()` outside of the `if` and then say `if (currentContents !== undefined) { ...` instead.